src/ITensorNetworks/peps.jl

```diff
@@ -0,0 +1,142 @@
+    end
+  end
+  lv = Matrix{Index}(undef, Ny - 1, Nx)
+  for ii in 1:(Nx)
+    for jj in 1:(Ny - 1)
+      lv[jj, ii] = Index(linkdims, "Lv,$jj,$ii")
+    end
+  end
+
+  # boundary cases
+  tensor_grid[1, 1] = emptyITensor(T, lh[1, 1], lv[1, 1], sites[1, 1])
+  tensor_grid[1, Nx] = emptyITensor(T, lh[1, Nx - 1], lv[1, Nx], sites[1, Nx])
+  tensor_grid[Ny, 1] = emptyITensor(T, lh[Ny, 1], lv[Ny - 1, 1], sites[Ny, 1])
+  tensor_grid[Ny, Nx] = emptyITensor(T, lh[Ny, Nx - 1], lv[Ny - 1, Nx], sites[Ny, Nx])
+  for ii in 2:(Nx - 1)
+    tensor_grid[1, ii] = emptyITensor(T, lh[1, ii], lh[1, ii - 1], lv[1, ii], sites[1, ii])
+    tensor_grid[Ny, ii] = emptyITensor(
+      T, lh[Ny, ii], lh[Ny, ii - 1], lv[Ny - 1, ii], sites[Ny, ii]
+    )
+  end
+
+  # inner sites
+  for jj in 2:(Ny - 1)
+    tensor_grid[jj, 1] = emptyITensor(T, lh[jj, 1], lv[jj, 1], lv[jj - 1, 1], sites[jj, 1])
+    tensor_grid[jj, Nx] = emptyITensor(
+      T, lh[jj, Nx - 1], lv[jj, Nx], lv[jj - 1, Nx], sites[jj, Nx]
+    )
+    for ii in 2:(Nx - 1)
+      tensor_grid[jj, ii] = emptyITensor(
+        T, lh[jj, ii], lh[jj, ii - 1], lv[jj, ii], lv[jj - 1, ii], sites[jj, ii]
+      )
+    end
+  end
+
+  return PEPS(tensor_grid)
+end
+
+PEPS(sites::Matrix{<:Index}, args...; kwargs...) = PEPS(Float64, sites, args...; kwargs...)
+
+PEPS(data::Array, size::Tuple{Int64,Int64}) = PEPS(reshape(data, size[1], size[2]))
+
+function randomizePEPS!(P::PEPS)
+  dimy, dimx = size(P.data)
+  for ii in 1:dimx
+    for jj in 1:dimy
+      randn!(P.data[jj, ii])
+      normalize!(P.data[jj, ii])
+    end
+  end
+end
+
+function Base.:+(A::PEPS, B::PEPS)
+  @assert(size(A.data) == size(B.data))
+  out = [t1 + t2 for (t1, t2) in zip(vcat(A.data...), vcat(B.data...))]
+  return PEPS(out, size(A.data))
+end
+
+function Base.:-(A::PEPS, B::PEPS)
+  @assert(size(A.data) == size(B.data))
+  out = [t1 - t2 for (t1, t2) in zip(vcat(A.data...), vcat(B.data...))]
+  return PEPS(out, size(A.data))
+end
+
+function Base.:*(c::Number, A::PEPS)
+  out = [c * t for t in vcat(A.data...)]
+  return PEPS(out, size(A.data))
+end
+
+function Base.:*(A::PEPS, B::PEPS)
+  @assert(size(A.data) == size(B.data))
+  out_scalar = 0
+  for (t1, t2) in zip(vcat(A.data...), vcat(B.data...))
+    out_scalar += ITensors.scalar(t1 * t2)
+  end
+  return out_scalar
+end
+
+get_prime_bonds(indices; ham=true) = ham ? indices : indices[1:(end - 1)]
+
+function ITensors.prime(peps::PEPS; ham=true)
+  prime_peps = map(
+    tensor -> prime(tensor, get_prime_bonds(inds(tensor); ham=ham)), peps.data
+  )
+  return PEPS(prime_peps)
+end
+
+# Get the tensor network of <peps|peps'>
+function inner_network(peps::PEPS, peps_prime::PEPS)
+  return vcat(vcat(peps.data...), vcat(peps_prime.data...))
+end
+
+# Get the tensor network of <peps|mpo|peps'>
+# The local MPO specifies the 2-site term of the Hamiltonian
+function inner_network(
+  peps::PEPS, peps_prime::PEPS, peps_prime_ham::PEPS, mpo::MPO, coordinates::Array
```

Comment:
Same question as above, couldn't this function do the priming internally?

Reply:
Also, is the current plan to optimize with only 2-site MPOs, or MPOs that extend across entire rows and columns? Either way is fine with me, whatever is easiest to get working at first. I'm just curious what you were thinking in terms of that.

Reply:
I plan to optimize MPOs across entire rows/columns. This should be easy to change later.